tests/pinmaplat.cpp

```diff
@@ -0,0 +1,207 @@
+                break;
+            case 'd':
+                dev_id = strtol(optarg, NULL, 0);
+                break;
+            case 'n':
+                num_iters = strtol(optarg, NULL, 0);
+                break;
+            case 'h':
+                printf("syntax: %s -s <buf size> -d <gpu dev id> -n <num iters> -h[help]\n", argv[0]);
+                exit(EXIT_FAILURE);
+                break;
+            default:
+                printf("ERROR: invalid option\n");
+                exit(EXIT_FAILURE);
+        }
+    }
+    
+    size_t size = (_size + GPU_PAGE_SIZE - 1) & GPU_PAGE_MASK;
+
+    ASSERTDRV(cuInit(0));
+
+    int n_devices = 0;
+    ASSERTDRV(cuDeviceGetCount(&n_devices));
+
+    CUdevice dev;
+    for (int n=0; n<n_devices; ++n) {
+        
+        char dev_name[256];
+        int dev_pci_domain_id;
+        int dev_pci_bus_id;
+        int dev_pci_device_id;
+
+        ASSERTDRV(cuDeviceGet(&dev, n));
+        ASSERTDRV(cuDeviceGetName(dev_name, sizeof(dev_name) / sizeof(dev_name[0]), dev));
+        ASSERTDRV(cuDeviceGetAttribute(&dev_pci_domain_id, CU_DEVICE_ATTRIBUTE_PCI_DOMAIN_ID, dev));
+        ASSERTDRV(cuDeviceGetAttribute(&dev_pci_bus_id, CU_DEVICE_ATTRIBUTE_PCI_BUS_ID, dev));
+        ASSERTDRV(cuDeviceGetAttribute(&dev_pci_device_id, CU_DEVICE_ATTRIBUTE_PCI_DEVICE_ID, dev));
+
+        cout  << "GPU id:" << n << "; name: " << dev_name 
+              << "; Bus id: "
+              << std::hex 
+              << std::setfill('0') << std::setw(4) << dev_pci_domain_id
+              << ":" << std::setfill('0') << std::setw(2) << dev_pci_bus_id
+              << ":" << std::setfill('0') << std::setw(2) << dev_pci_device_id
+              << std::dec
+              << endl;
+    }
+    cout << "selecting device " << dev_id << endl;
+    ASSERTDRV(cuDeviceGet(&dev, dev_id));
+
+    CUcontext dev_ctx;
+    ASSERTDRV(cuDevicePrimaryCtxRetain(&dev_ctx, dev));
+    ASSERTDRV(cuCtxSetCurrent(dev_ctx));
+
+    CUdeviceptr d_A;
+    ASSERTDRV(cuMemAlloc(&d_A, size));
+    cout << "device ptr: 0x" << hex << d_A << dec << endl;
+    cout << "allocated size: " << size << endl;
+
+    unsigned int flag = 1;
+    ASSERTDRV(cuPointerSetAttribute(&flag, CU_POINTER_ATTRIBUTE_SYNC_MEMOPS, d_A));
+
+    gdr_t g = gdr_open();
+    ASSERT_NEQ(g, (void*)0);
+
+    gdr_mh_t mh;
+    BEGIN_CHECK {
+        // tokens are optional in CUDA 6.0
+        // wave out the test if GPUDirectRDMA is not enabled
+
+        printf("Size(B) \t pin.Time(us) \t map.Time(us) \t get_info.Time(us) \t unmap.Time(us) \t unpin.Time(us)\n");
+        while (pin_request_size <= size) {
+            int iter = 0;
+            size_t actual_pin_size;
+            pin_lat_us = 0;
+            map_lat_us = 0;
+            unpin_lat_us = 0;
+            unmap_lat_us = 0;
+            inf_lat_us = 0;
+            actual_pin_size = (pin_request_size + GPU_PAGE_SIZE - 1) & GPU_PAGE_MASK;
+
+            for (iter = 0; iter < num_iters; ++iter) {
```

Comment:
Can you also add a warm up loop?

Reply:
maybe for a perf test, it is better to store all samples and to dump a histogram.
this would be related to workloads where a "registration cache" is not effective